### ISSUE TITLE
TIme 00:00

### ISSUE THREAD

**Issue by lubodrinka** [freeCodeCamp/testable-projects-fcc#376]
The test are not able to test of change time 00:00. I have special precision timer. My project work as is described in user stories.

#### Browser Information

* Browser Name, Version: all
* Operating System: windows10
* Mobile, Desktop, or Tablet: 

#### Your Code / Link to Your Pen

```
 https://codepen.io/vanderdrilu/pen/ZoVpPa/
 
 
```

#### Screenshot

Break Length
5
Session length
25
Pomodoro
FCC Test Suite: Pomodoro Clock
Select Test Suite: Run Tests Tests 20/29 Test(s) Failed  Tests Passed  Tests Executing Report Bug
Unit tests
passes: 20failures: 9duration: 31.90s
#Pomodoro Clock tests
#Technology Stack
1. You can use any mix of HTML, JavaScript, CSS, Bootstrap, SASS, React, Redux, and jQuery to complete this project. You should use a frontend framework (like React for example) because this section is about learning frontend frameworks. Additional technologies not listed above are not recommended and using them is at your own risk. We are looking at supporting other frontend frameworks like Angular and Vue, but they are not currently supported. We will accept and try to fix all issue reports that use the suggested technology stack for this project. Happy coding!
#Content
1. I can see an element with id="break-label" that contains a string (e.g. “Break Length”).
2. I can see an element with id="session-label" that contains a string (e.g. "Session Length”).
3. I can see two clickable elements with corresponding IDs: id="break-decrement" and id="session-decrement".
4. I can see two clickable elements with corresponding IDs: id="break-increment" and id="session-increment".
5. I can see an element, with corresponding id="break-length", which by default (on load) displays a value of 5.
6. I can see an element, with corresponding id="session-length", which by default displays a value of 25.
7. I can see an element, with corresponding id="timer-label", that contains a string indicating a session is initialized (e.g. "Session").
8. I can see an element with corresponding id="time-left". NOTE: Paused or running, the value in this field should always be displayed in mm:ss format (i.e. 25:00).
9. I can see a clickable element with corresponding id="start_stop".
10. I can see a clickable element with corresponding id="reset".
#Timer
1. When I click the element with the id of "reset", any running timer should be stopped, the value within id="break-length" should return to 5, the value within id="session-length" should return to 25, and the element with id="time-left" should reset to it's default state. Error: timeout of 5000ms exceeded. Ensure the done() callback is being called in this test.
    at https://cdnjs.cloudflare.com/ajax/libs/mocha/3.0.2/mocha.min.js:2:7197
2. When I click the element with the id of "break-decrement", the value within id="break-length" decrements by a value of 1, and I can see the updated value.
3. When I click the element with the id of "break-increment", the value within id="break-length" increments by a value of 1, and I can see the updated value.
4. When I click the element with the id of "session-decrement", the value within id="session-length" decrements by a value of 1, and I can see the updated value.
5. When I click the element with the id of "session-increment", the value within id="session-length" increments by a value of 1, and I can see the updated value.
6. I should not be able to set a session or break length to <= 0.67ms
7. I should not be able to set a session or break length to > 60.184ms
8. When I first click the element with id="start_stop", the timer should begin running from the value currently displayed in id="session-length", even if the value has been incremented or decremented from the original value of 25.
9. If the timer is running, the element with the id of "time-left" should display the remaining time in mm:ss format (decrementing by a value of 1 and updating the display every 1000ms). Error: Pomodoro has started but time displayed is not changing
    at https://cdn.freecodecamp.org/testable-projects-fcc/v1/bundle.js:20818:23
10. If the timer is running and I click the element with id="start_stop", the countdown should pause. Error: Pomodoro has started but time displayed is not changing
    at https://cdn.freecodecamp.org/testable-projects-fcc/v1/bundle.js:20835:23
11. If the timer is paused and I click the element with id="start_stop", the countdown should resume running from the point at which it was paused. Error: Pomodoro has started but time displayed is not changing
    at https://cdn.freecodecamp.org/testable-projects-fcc/v1/bundle.js:20864:23
12. When a session countdown reaches zero (NOTE: timer MUST reach 00:00), and a new countdown begins, the element with the id of "timer-label" should display a string indicating a break has begun. Error: timeout of 5000ms exceeded. Ensure the done() callback is being called in this test.
    at https://cdnjs.cloudflare.com/ajax/libs/mocha/3.0.2/mocha.min.js:2:7197
13. When a session countdown reaches zero (NOTE: timer MUST reach 00:00), a new break countdown should begin, counting down from the value currently displayed in the id="break-length" element. Error: timeout of 5000ms exceeded. Ensure the done() callback is being called in this test.
    at https://cdnjs.cloudflare.com/ajax/libs/mocha/3.0.2/mocha.min.js:2:7197
14. When a break countdown reaches zero (NOTE: timer MUST reach 00:00), and a new countdown begins, the element with the id of "timer-label" should display a string indicating a session has begun. Error: timeout of 5000ms exceeded. Ensure the done() callback is being called in this test.
    at https://cdnjs.cloudflare.com/ajax/libs/mocha/3.0.2/mocha.min.js:2:7197
15. When a break countdown reaches zero (NOTE: timer MUST reach 00:00), a new session countdown should begin, counting down from the value currently displayed in the id="session-length" element. Error: timeout of 5000ms exceeded. Ensure the done() callback is being called in this test.
    at https://cdnjs.cloudflare.com/ajax/libs/mocha/3.0.2/mocha.min.js:2:7197
#Audio
1. When a countdown reaches zero (NOTE: timer MUST reach 00:00), a sound indicating that time is up should play. This should utilize an HTML5 <audio> tag and have a corresponding id="beep". Error: timeout of 5000ms exceeded. Ensure the done() callback is being called in this test.
    at https://cdnjs.cloudflare.com/ajax/libs/mocha/3.0.2/mocha.min.js:2:7197
2. The audio element with id="beep" must be 1 second or longer.
3. The audio element with id of "beep" must stop playing and be rewound to the beginning when the element with the id of "reset" is clicked.
Close

**Comment by tbushman** [comment]
@lubodrinka Hello, I think your timer is accurate, and since you're using timestamps within the setInterval function, the duration of the interval can't be changed. The test suite speeds up all setInterval functions. You will need to decrement your timer by one second per interval regardless of current time in order to pass the tests.  
@tchaffee @no-stack-dub-sack What do you think about this case? Here's the main part that decrements @lubodrinka 's timer:  
```
    var date = new Date();
    var dateMS = date.getTime();
    var lastMS = last_time.getTime();
    var diff = dateMS - lastMS;
    // nastavený čas 25 min mínus rozdiel medzi dvomi časmi plus pauza
    diff = pausetimeMS + coundowntimeMs - diff;
    var outdate = new Date(diff);
    var min = outdate.getMinutes();
    var sec = outdate.getSeconds();
    var timeshort = (min < 10 ? '0' : '') + min + String.fromCharCode(58) + (sec < 10 ? '0' : '') + sec;
    $('#time-left').html(timeshort);
```   
where `#time-left` reflects actual elapsed time.

**Comment by tbushman** [comment]
@lubodrinka Sorry for the delay on this. For the purposes of these tests, please use either / both setInterval() / setTimeout() in order that the test can artificially speed up the timer. You have an accurate timer that uses timestamps, which is valid but the tests cannot accommodate them. Very sorry!

**Comment by tchaffee** [comment]
I'm closing this since we don't have any immediate plans to accommodate this way of timing at the moment. @lubodrinka Sorry about that - your code does look valid. We just can't make the tests work with this kind of code unless we completely change the way we test this project, and I'm not even sure how we would do that. If you are super interested you could take a look at our code and see if you can come up with any ideas. And please feel free to re-open this issue if there is still something you want to discuss about it.

**Comment by lubodrinka** [comment]
o.k I look later.

**Comment by lubodrinka** [comment]
and if you just set the time to 5 second and the test goes in real time?

**Comment by tbushman** [comment]
@lubodrinka Yes. Your timer uses real time / queries actual time whereas the tests use increments of variable time intervals.  
https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setInterval  

@tchaffee fyi, as @laniehei pointed out in #372 students are unable to reopen issues if they're closed by collaborators. :/

**Comment by tchaffee** [comment]
> students are unable to reopen issues if they're closed by collaborators.

@tbushman Thank you for letting me know about that. I'll word it differently going forward:

@lubodrinka If you want this issue reopened, just let us know.

**Comment by lubodrinka** [comment]
We don't understand to each together. I suppose the freecodcamp verificator could two version. In the slow version, the tester set time only once that worked, the test start in 00:59 instead 25:00, so what if for each test set 00:05 sec and test should pass.

**Comment by tbushman** [comment]
@lubodrinka This is a good consideration if we decide to accommodate different modes of timing in the future: To use short intervals for the tests and use unix date queries... We will have to keep it in mind and there's a short note about the timing hack in the documentation: `src/project-tests/pomodoro-clock-tests.js`. If you come up with a code schema for this, will it be ok to open a new issue?

**Comment by lubodrinka** [comment]
I look at file there should be  a preflight check if the timer has changed if not do not change to 30 ms and these two functions will be nested with function above
 function hackGlobalTimerFunctions() {
    window.setTimeout = (fun) => {
      return savedSetTimeout(fun, 30);
    };
    window.setInterval = (fun) => {
      return savedSetInterval(fun, 30);
    };
}
I try download the github put in pomodoro file
and set the path
<script src="testable-projects-fcc/build/testable-projects-fcc/v1/bundle.js"></script>in head 
but no test are shown on live server, how to handle that?

**Comment by tbushman** [comment]
@lubodrinka The problem with omitting `hackGlobalTimerFunctions()` for the tests is that the mocha timeouts would be reached before tests could complete. Each testable-projects test times out after at most ten seconds, so further changes would be required. If you want to run tests locally, follow the directions in https://github.com/freeCodeCamp/testable-projects-fcc/blob/master/CONTRIBUTING.md to get a local copy of the repo running on your machine. Changes like your suggestion will require code review with a pull request from a branch on your forked copy of this repo.  
If you have the repo forked, then the development bundle.js will become available in the `local_test` folder when you run `npm start` along with any project files you put in `local_test`

**Comment by lubodrinka** [comment]
mocha timeouts should be also variable increased by factor 1000/30

**Comment by lubodrinka** [comment]
I dont get it how to setup the local test.
The guide is incomplete.
I forked, downloaded and now what ? there main file testable-projects-fcc-master
so copy into pomodoro project directory or pomodoro files copy in` local_test?` directory 
I try run the yarn install in main testable-projects-fcc-master and in local_test directory pomodoro\testable-projects-fcc-master\local_test
is the yarn start the same as . code ?

**Comment by lubodrinka** [comment]
What is meaning by changes the index.html and than be sure to back to original state.
I create in `local_test` subdirectories css and js because in html is in lowercase.
```
PS C:\Users\lubo drinka\Documents\programovanie\jS ms\pomodoro\testable-projects-fcc-master\local_test> yarn start
yarn : The term 'yarn' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the
 spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ yarn start
+ ~~~~
    + CategoryInfo          : ObjectNotFound: (yarn:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

PS C:\Users\lubo drinka\Documents\programovanie\jS ms\pomodoro\testable-projects-fcc-master\local_test> yarn install
yarn : The term 'yarn' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the
 spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ yarn install
+ ~~~~
    + CategoryInfo          : ObjectNotFound: (yarn:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

PS C:\Users\lubo drinka\Documents\programovanie\jS ms\pomodoro\testable-projects-fcc-master\local_test>


``````
<!DOCTYPE html>
<html >
<head>
  <meta charset="UTF-8">
  <title>Local Test</title>
    
    <link rel="stylesheet" href="css/style.css">
    
</head>
<body>
  
  
  
  
  
  
  
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.1.0/jquery.min.js"></script> 
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.0.0/css/bootstrap.css">
<link href="https://fonts.googleapis.com/css?family=Monsieur+La+Doulaise" rel="stylesheet">
<script>
    src = "https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js"
</script>
<script>
    src = "https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.0.0/js/bootstrap.min.js"
</script>
  

  
  <script src='js/bundle.js'></script>

  
  <script src="js/pomodoro.js"></script>
</body>
</html>
```

**Comment by lubodrinka** [comment]
I get 
```
yarn start
yarn run v1.7.0
$ webpack --watch --config webpack.config.dev.js

Webpack is watching the files…
```

Hash: c7d45d89aac238148f08
Version: webpack 4.10.2
Time: 10201ms
Built at: 2018-07-12 14:35:17
    Asset      Size  Chunks             Chunk Names
bundle.js  1.07 MiB    main  [emitted]  main
[./node_modules/style-loader/index.js!./node_modules/css-loader/index.js!./src/stylesheets/fcc-test-toggler.css] ./node_modules/style-loader!./node_modules/css-loader!./src/stylesheets/fcc-test-toggler.css 1.02 KiB {main} [built]
[./node_modules/style-loader/index.js!./node_modules/css-loader/index.js!./src/stylesheets/fcc-test-ui.css] ./node_modules/style-loader!./node_modules/css-loader!./src/stylesheets/fcc-test-ui.css 1.01 KiB {main} [built]
[./node_modules/style-loader/index.js!./node_modules/css-loader/index.js!./src/stylesheets/mocha-modal.css] ./node_modules/style-loader!./node_modules/css-loader!./src/stylesheets/mocha-modal.css 1.01 KiB {main} [built]
[./src/index.js] 16.7 KiB {main} [built]
[./src/project-tests/portfolio-tests.js] 8.99 KiB {main} [built]
[./src/project-tests/product-landing-page-tests.js] 11.4 KiB {main} [built]
[./src/project-tests/quote-machine-tests.js] 6.55 KiB {main} [built]
[./src/project-tests/scatter-plot-tests.js] 8.93 KiB {main} [built]
[./src/project-tests/survey-form-tests.js] 12.3 KiB {main} [built]
[./src/project-tests/technical-docs-tests.js] 13.5 KiB {main} [built]
[./src/project-tests/tree-map-tests.js] 5.89 KiB {main} [built]
[./src/project-tests/tribute-page-tests.js] 6.29 KiB {main} [built]
[./src/utils/fcc-test-suite-skeleton.html] 2.14 KiB {main} [built]
[./src/utils/fcc-test-toggler-skeleton.html] 527 bytes {main} [built]
[./src/utils/mocha-modal-skeleton.html] 626 bytes {main} [built]
    + 62 hidden modules

```I use the index.html
 which is fcc
<!DOCTYPE html>
<html >
<head>
  <meta charset="UTF-8">
  <title>Local Test</title>
    <!-- ADD YOUR CSS HERE: local_test/css -->
    <link rel="stylesheet" href="css/style.css">
    <!-- <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css"> -->
</head>
<body>
  <!-- CHOOSE/ADD EXTERNAL RESOURCES HERE -->
  <!-- <script src='https://cdnjs.cloudflare.com/ajax/libs/react/15.4.2/react.min.js'></script> -->
  <!-- <script src='https://cdnjs.cloudflare.com/ajax/libs/react/15.4.2/react-dom.min.js'></script> -->
  <!-- <script src='https://d3js.org/d3.v4.min.js'></script> -->
  <!-- <script src='https://d3js.org/topojson.v2.min.js'></script> -->
  <!-- <script src='https://d3js.org/d3-scale-chromatic.v1.min.js'></script> -->
  <!-- <script src="https://cdnjs.cloudflare.com/ajax/libs/marked/0.3.6/marked.js"></script> -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.1.0/jquery.min.js"></script> 


  <!-- <div id="app"></div> -->

  <!-- KEEP THIS AND MAKE SURE YOU'VE CREATED A local_test/js directory -->
  <script src='js/bundle.js'></script>

  <!-- ADD YOUR LOCAL JS HERE: local_test/js -->
  <script src="js/pomodoro.js"></script>
</body>
</html>
```

**Comment by lubodrinka** [comment]
testable-projects-fcc-master\local_test> yarn serve-https-local
yarn run v1.7.0
error Command "serve-https-local" not found.

**Comment by lubodrinka** [comment]
this guide is bad
I finally try put down   <script src='/testable-projects-fcc-master/local_test/js/bundle.js'></script> into my pomodoro html and it worked

**Comment by lubodrinka** [comment]
function getSessionlengtElementhvalue() {
  return document.getElementById('session-length');
}
  this.timeout(getSessionlengtElementhvalue() + 5000);

**Comment by tbushman** [comment]
OK. Sorry for the trouble.

**Comment by lubodrinka** [comment]
I found there is some, contrary goals the test should start in full time for example 1:00,but if the timer start so, the  others goals which say it must in from session to break in  haven't pass. and also that conflict somehow with audio, because 2. second goals is now not  fulfill. If add ne sec in start time other which control the text in session-length are not pass.

**Comment by tchaffee** [comment]
> this guide is bad

@lubodrinka  we will also consider PRs that improve the guide if you have ideas about how to make it clearer.
